File: outco/curriculum/homework_solutions/java/w3_d3_binary_heap.java
 /*
  * Homework  - Heap
  *
  *
  *  Prompt: Create a Heap class/constructor
  *
  *  The Heap will take in the following input:
  *
  *                type: argument should be either 'min' or 'max'. This will
  *                      dictate whether the heap will be a minheap or maxheap.
  *                      The comparator method will be affected by this
  *                      argument. See method description below
  *
  *  The Heap will have the following property:
  *
  *             storage: array
  *
  *                type: property that will be either 'min' or 'max'. This will
  *                      be dictated by
  *
  *  The Heap will have the following methods:
  *
  *             compare: takes in two arguments (a and b). Depending on the heap
  *                      type (minheap or maxheap), the comparator will behave
  *                      differently. If the heap is a minheap, then the compare
  *                      function will return true if a is less than b, false
  *                      otherwise. If the heap is a maxheap, then the compare
  *                      function will return true if a is greater than b, false
  *                      otherwise.
  *
  *                swap: takes in two indexes and swaps the elements at the two
  *                      indexes in the storage array
  *
  *                peak: returns the peak element of the storage array. This
  *                      will be the most minimum and maximum element for a
  *                      minheap and maxheap respectively
  *
  *                size: returns the number of elements in the heap
  *
  *              insert: inserts a value into the heap. Should begin by pushing
  *                      the value onto the end of the array, and then calling
  *                      the bubbleUp method from the last index of the array
  *
  *            bubbleUp: takes in an index, and considers the element at that
  *                      index to be a child. Continues to swap that child with
  *                      its parent value if the heap comparator condition is
  *                      not fulfilled
  *
  *          removePeak: removes the peak element from the heap and returns it.
  *                      Should begin by swapping the 0th-index element with the
  *                      last element in the storage array. Uses bubbleDown
  *                      method from the 0th index
  *
  *          bubbleDown: takes in an index, and considers the element at this
  *                      index to be the parent. Continues to swap this parent
  *                      element with its children if the heap comparator
  *                      condition is not fulfilled
  *
  *              remove: takes in a value and (if it exists in the heap),
  *                      removes that value from the heap data structure and
  *                      returns it. Should rearrange the rest of the heap to
  *                      ensure the heap comparator conditions are fulfilled
  *                      for all of its elements
  *
  *
  *
  *  Input:  N/A
  *  Output: A Heap instance
  *
  *  What are the time and auxiliary space complexities of the various methods?
  *
  */

import java.util.*;

class Heap {

  List<Integer> storage;
  String type;


  public Heap(String type) {
    this.type = type;
    this.storage = new ArrayList<>();
  }

  // Time Complexity: O(1)
  // Auxiliary Space Complexity: O(1)
  public boolean compare(int a, int b){
    if (type.equals("max")) {
      return storage.get(a) >= storage.get(b);
    } else {
      return storage.get(a) <= storage.get(b);
    }
  }

  // Time Complexity: O(1)
  // Auxiliary Space Complexity: O(1)
  public void swap(int index1, int index2){
    Collections.swap(this.storage, index1, index2);
  }

  // Time Complexity: O(1)
  // Auxiliary Space Complexity: O(1)
  public int peak(){
    if(this.size() == 0) return -1;
    return storage.get(0);
  }

  // Time Complexity: O(1)
  // Auxiliary Space Complexity: O(1)
  public int size(){
    return storage.size();
  }

  // Time Complexity: O(log(N))
  // Auxiliary Space Complexity: O(1)
  public void insert(int value){
    storage.add(value);
    bubbleUp(this.size() - 1);
  }

  // Time Complexity: O(log(N))
  // Auxiliary Space Complexity: O(1)
  public void bubbleUp(int childIndex){
    if (childIndex < 1 || childIndex >= this.size()) { return; }

    int parentIndex = (int) Math.floor((childIndex-1.0)/2.0);

    while (childIndex > 0 && !compare(parentIndex, childIndex)) {
      swap(parentIndex, childIndex);
      childIndex = parentIndex;
      parentIndex = (int) Math.floor((childIndex-1.0)/2.0);
    }
  }

  // Time Complexity: O(log(N))
  // Auxiliary Space Complexity: O(1)
  public int removePeak(){
    swap(0, size() - 1);
    int toReturn = storage.remove(size()-1);
    bubbleDown(0);
    return toReturn;
  }


  private int _getChild(int parentIndex) {
    int child1Index = 2 * parentIndex + 1;
    int child2Index = 2 * parentIndex + 2;

    if (child1Index >= this.size() || child2Index >= this.size() || this.compare(child1Index, child2Index)){
      return child1Index;
    } else {
      return child2Index;
    }
  }

  // Time Complexity: O(log(N))
  // Auxiliary Space Complexity: O(1)
  public void bubbleDown(int parentIndex){
    if (parentIndex >= this.size()) { return;  }

    int childIndex = _getChild(parentIndex);

    while (childIndex < this.size() && !compare(parentIndex, childIndex)) {
      swap(parentIndex, childIndex);
      parentIndex = childIndex;
      childIndex = _getChild(parentIndex);
    }
  }

  /*To implement remove, iterate through the array until you find the index
  where the value to be removed is located. When found, swap this element
  with the last element of the storage array, and then pop it off. Then run
  bubbleUp and bubbleDown from the index where the last index was swapped in.*/
  // Time Complexity: O(N)
  // Auxiliary Space Complexity: O(1)
  public boolean remove(int value){
    for (int i = 0 ; i < this.size() ; i++) {
      if (storage.get(i) == value) {
        swap(i, this.size() - 1);
        storage.remove(size()-1);
        bubbleUp(i);
        bubbleDown(i);
        return true;
      }
    }
    return false;
  }
}




////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

// use the Main class to run the test cases
class HeapTests {
  // an interface to perform tests
  public interface Test {
    boolean execute();
  }

  public static void main(String[] args) {

    // instantiate the testing of each module by resetting count and printing title of module
    int[] testCount = {0, 0};
    System.out.println("Heap Class");

    // tests are in the form as shown
    assertTest(testCount, "has storage field", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getDeclaredField("storage");
        return true;
      } catch (Exception e) {
        return false;
      }
    });

    assertTest(testCount, "has type field", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getDeclaredField("type");
        return true;
      } catch (Exception e) {
        return false;
      }
    });

    assertTest(testCount, "default storage set to an array", () -> {
      Heap heap = new Heap("max");
      try {
        return heap.getClass().getDeclaredField("storage").getType().isAssignableFrom(List.class) ||
                heap.getClass().getDeclaredField("storage").getType().isAssignableFrom(ArrayList.class);
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "default type property is set to 'min'", () -> {
      Heap heap = new Heap("min");
      return heap.type.equals("min");
    });

    assertTest(testCount, "default type property is set to 'max'", () -> {
      Heap heap = new Heap("max");
      return heap.type.equals("max");
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap compare method");

    // tests are in the form as shown
    assertTest(testCount, "has compare method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("compare", int.class, int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "returns true for minheap if element at first argument index is less than element at second argument index", () -> {
      Heap heap = new Heap("min");
      heap.storage.add(1);
      heap.storage.add(10);
      return heap.compare(0, 1);
    });

    assertTest(testCount, "returns false for minheap if element at first argument index is greater than element at second argument index", () -> {
      Heap heap = new Heap("min");
      heap.storage.add(10);
      heap.storage.add(1);
      return !heap.compare(0, 1);
    });

    assertTest(testCount, "return true for maxheap if element at first argument index is greater than element at second argument index", () -> {
      Heap heap = new Heap("max");
      heap.storage.add(10);
      heap.storage.add(1);
      return heap.compare(0, 1);
    });

    assertTest(testCount, "return false for maxheap if element at first argument index is less than element at second argument index", () -> {
      Heap heap = new Heap("max");
      heap.storage.add(1);
      heap.storage.add(10);
      return !heap.compare(0, 1);
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap swap method");

    // tests are in the form as shown
    assertTest(testCount, "has swap method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("swap", int.class, int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should be able to swap the locations of two elements given two indices", () -> {
      Heap heap = new Heap("min");
      heap.storage.add(1);
      heap.storage.add(5);
      heap.storage.add(10);
      heap.swap(0, 2);
      return heap.storage.get(0) == 10 && heap.storage.get(2) == 1;
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap peak method");

    // tests are in the form as shown
    assertTest(testCount, "has peak method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("peak");
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should return the 0th element of the storage array", () -> {
      Heap heap = new Heap("min");
      heap.storage.add(1);
      heap.storage.add(5);
      heap.storage.add(10);
      return heap.peak() == 1;
    });

    assertTest(testCount, "upon building out your insert method, should return the smallest element for a minheap", () -> {
      Heap heap = new Heap("min");
      heap.insert(2);
      heap.insert(5);
      heap.insert(10);
      heap.insert(1);
      return heap.peak() == 1;
    });

    assertTest(testCount, "upon building out your insert method, should return the largest element for a maxheap", () -> {
      Heap heap = new Heap("max");
      heap.insert(2);
      heap.insert(5);
      heap.insert(10);
      heap.insert(1);
      return heap.peak() == 10;
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap size method");

    // tests are in the form as shown
    assertTest(testCount, "has size method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("size");
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "returns number of elements in the storage array", () -> {
      Heap heap = new Heap("min");
      heap.storage.add(1);
      heap.storage.add(5);
      heap.storage.add(10);
      return heap.size() == 3;
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap insert method");

    // tests are in the form as shown
    assertTest(testCount, "has insert method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("insert", int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should be able to insert a single element", () -> {
      Heap heap = new Heap("min");
      heap.insert(5);
      return heap.storage.get(0) == 5;
    });

    assertTest(testCount, "should be able to insert multiple elements into a minheap and have peak element be smallest element", () -> {
      Heap heap = new Heap("min");
      heap.insert(5);
      heap.insert(10);
      heap.insert(7);
      heap.insert(1);
      heap.insert(8);
      heap.insert(6);
      return heap.peak() == 1;
    });

    assertTest(testCount, "should be able to insert multiple elements into a maxheap and have peak element be largest element", () -> {
      Heap heap = new Heap("max");
      heap.insert(5);
      heap.insert(10);
      heap.insert(7);
      heap.insert(1);
      heap.insert(8);
      heap.insert(6);
      return heap.peak() == 10;
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap bubbleUp method");

    // tests are in the form as shown
    assertTest(testCount, "has bubbleUp method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("bubbleUp", int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should be able to 'bubble up' an element if its minheap condition is not fulfilled", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(2,4,7,6,9,10,8,1));
      heap.bubbleUp(7);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6)));
    });

    assertTest(testCount, "should be able to 'bubble up' an element if its maxheap condition is not fulfilled", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(9,8,6,7,3,5,2,10));
      heap.bubbleUp(7);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7)));
    });

    assertTest(testCount, "should not perform bubbling up if minheap conditions are fulfilled", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      heap.bubbleUp(7);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6)));
    });

    assertTest(testCount, "should not perform bubbling up if maxheap conditions are fulfilled", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      heap.bubbleUp(7);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7)));
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap removePeak method");

    // tests are in the form as shown
    assertTest(testCount, "has removePeak method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("removePeak");
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should be able to remove a single element", () -> {
      Heap heap = new Heap("min");
      heap.insert(5);
      heap.removePeak();
      return heap.storage.size() == 0;
    });

    assertTest(testCount, "should be able to remove and return peak element for a minheap and rearrange minheap accordingly", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      int peak = heap.removePeak();
      return peak == 1 && heap.storage.equals(new ArrayList<>(Arrays.asList(2,4,7,6,9,10,8)));
    });

    assertTest(testCount, "should be able to remove and return peak element for a maxheap and rearrange maxheap accordingly", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      int peak = heap.removePeak();
      return peak == 10 && heap.storage.equals(new ArrayList<>(Arrays.asList(9,8,6,7,3,5,2)));
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap bubbleDown method");

    // tests are in the form as shown
    assertTest(testCount, "has bubbleDown method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("bubbleDown", int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "should be able to 'bubble down' an element if its minheap condition is not fulfilled", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(10,1,2,6,4,9,8,7));
      heap.bubbleDown(0);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(1,4,2,6,10,9,8,7)));
    });

    assertTest(testCount, "should be able to 'bubble down' an element if its maxheap condition is not fulfilled", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(2,10,9,5,8,3,6,7));
      heap.bubbleDown(0);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(10,8,9,5,2,3,6,7)));
    });

    assertTest(testCount, "should not perform bubbling down if minheap conditions are fulfilled", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      heap.bubbleDown(0);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6)));
    });

    assertTest(testCount, "should not perform bubbling down if maxheap conditions are fulfilled", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      heap.bubbleDown(0);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7)));
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



    // instantiate the testing of each module by resetting count and printing title of module
    testCount[0] = 0;
    testCount[1] = 0;
    System.out.println("Heap remove method");

    // tests are in the form as shown
    assertTest(testCount, "has remove method", () -> {
      Heap heap = new Heap("max");
      try {
        heap.getClass().getMethod("remove", int.class);
        return true;
      } catch (Exception e) {
        e.printStackTrace();
        return false;
      }
    });

    assertTest(testCount, "is able to remove specified value from minheap", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      boolean removed = heap.remove(10);
      return removed && heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,6,4,9,7,8)));
    });

    assertTest(testCount, "is able to remove specified value from maxheap", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      boolean removed = heap.remove(6);
      return removed && heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,7,8,3,5,2)));
    });

    assertTest(testCount, "is able to remove last value from minheap", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      boolean removed = heap.remove(6);
      return removed && heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8)));
    });

    assertTest(testCount, "is able to remove last value from maxheap", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      boolean removed = heap.remove(7);
      return removed && heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2)));
    });

    assertTest(testCount, "does not remove anything from minheap if value does not exist", () -> {
      Heap heap = new Heap("min");
      heap.storage = new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6));
      heap.remove(3);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(1,2,7,4,9,10,8,6)));
    });

    assertTest(testCount, "does not remove anything from maxheap if value does not exist", () -> {
      Heap heap = new Heap("max");
      heap.storage = new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7));
      heap.remove(4);
      return heap.storage.equals(new ArrayList<>(Arrays.asList(10,9,6,8,3,5,2,7)));
    });

    // print the result of tests passed for a module
    System.out.println("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");



  }

  // do not edit below, this is to wrap the test and check for exceptions
  private static void assertTest(int[] count, String name, Test test) {
    String pass = "false";
    count[1]++;

    try {
      if (test.execute()) {
        pass = " true";
        count[0]++;
      }
    } catch(Exception ignored) {}
    String result = "  " + (count[1] + ")   ").substring(0, 5) + pass + " : " + name;
    System.out.println(result);
  }
}

File: outco/curriculum/homework_solutions/swift/w3_d3_binary_heap.swift
/**
 *  Homework - Binary Heap
 *
 *  Prompt: Create a Binary Heap class
 *
 *  The Heap will take in the following input:
 *
 *            type:   String - 'min' for minheap, 'max' for maxheap
 *
 *  The Heap will have the following properties:
 *
 *         storage:   [Int] - storage of integers within heap
 *            type:   String - idicates whether heap is a minheap or maxheap
 *
 *  The Heap will have the following methods:
 *
 *        compare:   compares two input numbers and returns if the heap
 *                   condition is met.
 *
 *                   Input:      parent: Int - index at parent
 *                   Input:      child: Int - index at child
 *                   Output:     Bool
 *
 *         insert:   inserts a number into the heap.
 *
 *                    Input:     Int
 *                   Output:     Void
 *
 *           peek:   returns the top item in the heap but does not remove it
 *
 *                   Input:      N/A
 *                   Output:     Int?
 *
 *         removePeak:   removes the top element from the heap and returns it
 *
 *                    Input:     N/A
 *                   Output:     Int?
 */

class Heap {
  var storage = [Int]()
  var type: String

  init(_ type: String) {
    self.type = type == "min" ? "min" : "max"
  }

  init() {
    self.type = "min"
  }

  // Time Complexity: O(1)
  // Auxiliary Space Complexity: O(1)
  func compare(_ parent: Int, _ child: Int) -> Bool {
   if type == "min" { return storage[parent] <= storage[child] }
   return storage[parent] > storage[child]
 }

 // Time Complexity: O(log(N))
 // Auxiliary Space Complexity: O(1)
 func insert(_ value: Int) {
   storage.append(value)
   bubbleUp()
 }


 // Time Complexity: O(1)
 // Auxiliary Space Complexity: O(1)
 func peek() -> Int? {
   return storage.count > 0 ? storage[0] : nil
 }

 // Time Complexity: O(log(N))
 // Auxiliary Space Complexity: O(1)
 func removePeak() -> Int? {
   if (storage.count == 0) { return nil }
   swap(0, storage.count - 1)
   let result = storage.removeLast()
   bubbleDown()
   return result
 }

 // Helper methods
 func swap(_ a: Int, _ b: Int) {
   (storage[a], storage[b]) = (storage[b], storage[a])
 }

 func getParent(_ child: Int) -> Int {
   return (child - 1) / 2
 }

 func getChild(_ parent: Int) -> Int {
   let left: Int = parent * 2 + 1;
   if (left >= storage.count - 1 || compare(left, left + 1)) {
     return left;
   }
   return left + 1;
 }

 func bubbleUp() {
   var child: Int = storage.count - 1
   var parent: Int = getParent(child)

   while child > 0 && !compare(parent, child) {
     swap(parent, child)
     child = parent
     parent = getParent(child)
   }
 }

 func bubbleDown() {
   var parent: Int = 0
   var child: Int = getChild(parent)

   while child < storage.count && !compare(parent, child) {
     swap(parent, child)
     parent = child
     child = getChild(parent)
   }
 }
}




////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////


func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}


print("Heap Class")
var testCount = [0, 0]

assert(&testCount, "has storage property",  {
  let heap = Heap()
  return heap.storage == [Int]()
})

assert(&testCount, "has type property min by default",  {
  let heap = Heap()
  return heap.type == "min"
})

assert(&testCount, "can set type property to max",  {
  let heap = Heap("max")
  return heap.type == "max"
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Heap compare method")
testCount = [0, 0]

assert(&testCount, "returns true for minheap if element at first argument index is less than element at second argument index", {
 let heap = Heap()
 heap.storage.append(1)
 heap.storage.append(10)
 return heap.compare(0, 1) == true
})

assert(&testCount, "returns false for minheap if element at first argument index is greater than element at second argument index", {
 let heap = Heap()
 heap.storage.append(10)
 heap.storage.append(1)
 return heap.compare(0, 1) == false
})

assert(&testCount, "return true for maxheap if element at first argument index is greater than element at second argument index", {
 let heap = Heap("max")
 heap.storage.append(10)
 heap.storage.append(1)
 return heap.compare(0, 1) == true
})

assert(&testCount, "return false for maxheap if element at first argument index is less than element at second argument index", {
 let heap = Heap("max")
 heap.storage.append(1)
 heap.storage.append(10)
 return heap.compare(0, 1) == false
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Heap insert method")
testCount = [0, 0]

assert(&testCount, "should be able to insert a single element", {
 let heap = Heap()
 heap.insert(5)
 return heap.storage.count > 0 && heap.storage[0] == 5
})

assert(&testCount, "minimum value should be on top of a minheap", {
 let heap = Heap()
 heap.insert(5)
 heap.insert(10)
 heap.insert(7)
 heap.insert(1)
 heap.insert(8)
 heap.insert(6)
 return heap.storage.count > 0 && heap.storage[0] == 1
})

assert(&testCount, "maximum value should be on top of a maxheap", {
 let heap = Heap("max")
 heap.insert(5)
 heap.insert(10)
 heap.insert(7)
 heap.insert(1)
 heap.insert(8)
 heap.insert(6)
 return heap.storage.count > 0 && heap.storage[0] == 10
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Heap peek method")
testCount = [0, 0]

assert(&testCount, "should return the top element of the heap", {
 let heap = Heap()
 heap.storage.append(1)
 return heap.peek()! == 1
})

assert(&testCount, "should return nil if the heap is empty", {
 let heap = Heap()
 return heap.peek() == nil
})

assert(&testCount, "should return the smallest element for a minheap", {
 let heap = Heap()
 heap.insert(2)
 heap.insert(5)
 heap.insert(10)
 heap.insert(1)
 return heap.peek() == 1
})

assert(&testCount, "should return the largest element for a maxheap", {
 let heap = Heap("max")
 heap.insert(2)
 heap.insert(5)
 heap.insert(10)
 heap.insert(1)
 return heap.peek() == 10
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Heap removePeak method")
testCount = [0, 0]

assert(&testCount, "should be able to remove a single element", {
 let heap = Heap()
 heap.insert(5)
 let min = heap.removePeak()
 return heap.storage.count == 0 && min == 5
})

assert(&testCount, "should return nil if the heap is empty", {
 let heap = Heap()
 return heap.removePeak() == nil
})

assert(&testCount, "should be able to remove and return min", {
 let heap = Heap("min")
 heap.storage = [1, 2, 7, 4, 9, 10, 8, 6]
 let min = heap.removePeak()
 return min == 1 && heap.peek() == 2
})

assert(&testCount, "should be able to remove and return max", {
 let heap = Heap("max")
 heap.storage = [10, 9, 6, 8, 3, 5, 2, 7]
 let max = heap.removePeak()
 return max == 10 && heap.peek() == 9
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
